Source/cairo/WaylandCairoSurface.m

```diff
@@ -0,0 +1,204 @@
+create_shm_buffer(struct window *window)
+{
+    struct wl_shm_pool *pool;
+    cairo_surface_t *surface;
+    int fd, size, stride;
+
+    stride = window->width * 4;
+    size = stride * window->height;
+
+    NSDebugLog(@"WaylandCairoSurface: creating shm buffer of %d bytes", size);
+    fd = os_create_anonymous_file(size);
+    if (fd < 0) {
+	NSLog(@"creating a buffer file for surface failed");
+	return NULL;
+    }
+
+    window->data =
+	mmap(NULL, size, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
+    if (window->data == MAP_FAILED) {
+	NSLog(@"error mapping anonymous file");
+	close(fd);
+	return NULL;
+    }
+
+    pool = wl_shm_create_pool(window->wlconfig->shm, fd, size);
+
+    surface = cairo_image_surface_create_for_data(window->data,
+						  CAIRO_FORMAT_ARGB32,
+						  window->width,
+						  window->height,
+						  stride);
+
+    window->buffer =
+	wl_shm_pool_create_buffer(pool, 0,
+				  window->width, window->height, stride,
+				  WL_SHM_FORMAT_ARGB8888);
+    wl_shm_pool_destroy(pool);
+
+    close(fd);
+
+    return surface;
+}
+
+@implementation WaylandCairoSurface
+
+- (id) initWithDevice: (void*)device
+{
+    struct window *window = (struct window *) device;
+    NSDebugLog(@"WaylandCairoSurface: initWithDevice win=%d", window->window_id);
+
+    gsDevice = device;
+
+    _surface = create_shm_buffer(window);
+    if (_surface == NULL) {
+	NSDebugLog(@"can't create cairo surface");
+	return 0;
+    }
+
+    wl_surface_attach(window->surface, window->buffer, 0, 0);
+    window->wcs = self;
+
+    return self;
+}
+
+- (void) dealloc
+{
+    struct window *window = (struct window*) gsDevice;
+    NSDebugLog(@"WaylandCairoSurface: dealloc win=%d", window->window_id);
+    [super dealloc];
+}
+
+- (NSSize) size
+{
+    NSDebugLog(@"WaylandCairoSurface: size");
+    struct window *window = (struct window*) gsDevice;
+    return NSMakeSize(window->width, window->height);
+}
+
+- (void) setSurface: (cairo_surface_t*)surface
+{
+    NSDebugLog(@"WaylandCairoSurface: setSurface");
+    _surface = surface;
+}
+
+- (void) handleExposeRect: (NSRect)rect
+{
+    NSDebugLog(@"handleExposeRect");
+    struct window *window = (struct window*) gsDevice;
+    cairo_surface_t *cairo_surface = _surface;
+    double  backupOffsetX = 0;
+    double  backupOffsetY = 0;
+    int x = NSMinX(rect);
+    int y = NSMinY(rect);
+    int width = NSWidth(rect);
+    int height = NSHeight(rect);
+
+    NSDebugLog(@"updating region: %dx%d %dx%d", x, y, width, height);
+
+    if (cairo_surface_status(cairo_surface) != CAIRO_STATUS_SUCCESS)
+    {
+	NSWarnMLog(@"cairo initial window error status: %s\n",
+		   cairo_status_to_string(cairo_surface_status(_surface)));
+    }
+
+    cairo_surface_get_device_offset(cairo_surface, &backupOffsetX, &backupOffsetY);
+    cairo_surface_set_device_offset(cairo_surface, 0, 0);
+
+    cairo_t *cr = cairo_create(cairo_surface);
+    if (width != window->width && 0) {
+	NSDebugLog(@"fake drawing");
```

Comment:
What is this branch for?

Reply:
I can't tell, and it was disabled in the patches we received, so I removed it. Presumably stub drawing code left-over from the original author.